src/pdn/src/PdnGen.cc

```diff
@@ -225,14 +251,629 @@ PdnGen::addGlobalConnect(dbBox* region, const char* instPattern, const char* pin
+                                     Rings::Layer{layer1, width1, spacing1}};
+  auto ring = std::make_unique<Rings>(grid, layers);
+  ring->setOffset(offset);
+  if (std::any_of(
+          pad_offset.begin(), pad_offset.end(), [](int o) { return o != 0; })) {
+    ring->setPadOffset(pad_offset);
+  }
+  ring->setExtendToBoundary(extend);
+  if (starts_with != GRID) {
+    ring->setStartWithPower(starts_with == POWER);
+  }
+  grid->addRing(std::move(ring));
+  if (!pad_pin_layers.empty() && grid->type() == Grid::Core) {
+    auto* core_grid = static_cast<CoreGrid*>(grid);
+    core_grid->setupDirectConnect(pad_pin_layers);
+  }
+  rendererRedraw();
+}
+
+void PdnGen::makeFollowpin(Grid* grid,
+                           odb::dbTechLayer* layer,
+                           int width,
+                           ExtensionMode extend)
+{
+  auto strap = std::make_unique<FollowPins>(grid, layer, width);
+  strap->setExtend(extend);
+
+  grid->addStrap(std::move(strap));
+  rendererRedraw();
+}
+
+void PdnGen::makeStrap(Grid* grid,
+                       odb::dbTechLayer* layer,
+                       int width,
+                       int spacing,
+                       int pitch,
+                       int offset,
+                       int number_of_straps,
+                       bool snap,
+                       StartsWith starts_with,
+                       ExtensionMode extend)
+{
+  auto strap = std::make_unique<Straps>(
+      grid, layer, width, pitch, spacing, number_of_straps);
+  strap->setExtend(extend);
+  strap->setOffset(offset);
+  strap->setSnapToGrid(snap);
+  if (starts_with != GRID) {
+    strap->setStartWithPower(starts_with == POWER);
+  }
+  grid->addStrap(std::move(strap));
+  rendererRedraw();
+}
+
+void PdnGen::makeConnect(Grid* grid,
+                         odb::dbTechLayer* layer0,
+                         odb::dbTechLayer* layer1,
+                         int cut_pitch_x,
+                         int cut_pitch_y,
+                         const std::vector<odb::dbTechViaGenerateRule*>& vias,
+                         const std::vector<odb::dbTechVia*>& techvias,
+                         int max_rows,
+                         int max_columns,
+                         const std::vector<odb::dbTechLayer*>& ongrid,
+                         const std::map<odb::dbTechLayer*, int>& split_cuts,
+                         const std::string& dont_use_vias)
+{
+  auto con = std::make_unique<Connect>(grid, layer0, layer1);
+  con->setCutPitch(cut_pitch_x, cut_pitch_y);
+
+  for (auto* via : vias) {
+    con->addFixedVia(via);
+  }
+
+  for (auto* via : techvias) {
+    con->addFixedVia(via);
+  }
+
+  con->setMaxRows(max_rows);
+  con->setMaxColumns(max_columns);
+  con->setOnGrid(ongrid);
+  con->setSplitCuts(split_cuts);
+
+  if (!dont_use_vias.empty()) {
+    con->filterVias(dont_use_vias);
+  }
+
+  grid->addConnect(std::move(con));
+  rendererRedraw();
+}
+
+void PdnGen::toggleDebugRenderer(bool on)
```

Comment:
enableDebugRenderer as this doesn't really toggle

Reply:
renamed and corrected some behavior